spring-kafka/src/main/java/org/springframework/kafka/core/reactive/ReactiveKafkaConsumerTemplate.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core.reactive;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.kafka.receiver.KafkaReceiver;
+import reactor.kafka.receiver.ReceiverOptions;
+import reactor.kafka.receiver.ReceiverRecord;
+import reactor.kafka.sender.TransactionManager;
+
+/**
+ * Reactive kafka consumer operations implementation.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ *
+ * @author Mark Norkin
+ */
+public class ReactiveKafkaConsumerTemplate<K, V> implements ReactiveKafkaConsumerOperations<K, V> {
+	private final KafkaReceiver<K, V> kafkaReceiver;
+
+	public ReactiveKafkaConsumerTemplate(ReceiverOptions<K, V> receiverOptions) {
+		this.kafkaReceiver = KafkaReceiver.create(receiverOptions);
+	}
+
+	@Override
+	public Flux<ReceiverRecord<K, V>> receive() {
+		return this.kafkaReceiver.receive();
+	}
+
+	@Override
+	public Flux<Flux<ConsumerRecord<K, V>>> receiveAutoAck() {
```

Comment:
I wonder how this `Flux<Flux<?>>` is useful on practice. Can't we just `flatMap()` it before returning?
Seems like API is claiming an `autoAck`, so, it doesn't bring me too much value to worry about and underneath "hard" logic.

Am I missing anything?

Reply:
Yes, this would simplify the api a bit.
However we have to assume that client is not interest in receiving each batch separately as stated in KafkaReceiver doc 
> Returns a {@link Flux} containing each batch of consumer records returned by {@link Consumer#poll(long)}

Do you know a use case which requires proccessing each autoAck batch separately, if so should we give the client an option then ?